proposals/0000-linear-types.rst

```diff
@@ -0,0 +1,2245 @@
+the exact same code: one for ``p=1`` and one for ``p=ω``. Function
+composition is even worse: it takes two multiplicity parameters,
+hence, would require four identical implementations:
+
+::
+
+  (.) :: (b :p-> c) -> (a :q-> b) -> a :(p ':* q)-> c
+
+Syntax
+------
+
+.. _Syntax:
+
+*This section is only a proposal for a syntax for linear types. We had
+to choose a syntax for the purpose of the proposal, but we are not
+particularly decided on it. We have listed all the alternatives in the
+Alternatives_ section.*
+
+This proposal adds two new syntactical constructs:
+
+- The multiplicity annotated arrow, for polymorphism, is written
+  ``a :p-> b`` (where ``a`` and ``b`` are types and ``p`` is a
+  multiplicity). To avoid introducing a new notion of "mixfix"
+  operators, we introduce a familly of (infix) type constructors:
+  ``(:p->)`` for each multiplicity ``p``. This technically steals
+  syntax as ``(:)`` is a valid type operator under the discouraged
+  ``-XDataKinds`` syntax. But this should not be a problem in
+  practice.
+
+  - In ``(:p->)``, ``p`` can be any type expression of kind
+    ``Multiplicity`` (see below). So that the following is legal
+    (though see Alternatives_):
+
+    ::
+
+      type familly F (a :: *) :: Multiplicity
+      f ::  forall (a :: *). Int  :(F a)-> a -> a
+- Binders can also be annotated with a multiplicity:
+
+  ::
+
+    \x ::('One) A -> x
+
+  is the identity function at type ``A ->. A``. This adds a new syntactic for binders:
+
+  ::
+
+    VAR ::( TYPE_EXPRESSION ) TYPE_EXPRESSION
+
+  where the first ``TYPE_EXPRESSION`` must be of kind ``Multiplicity``
+  (see below). This creates a new lexical token ``::(``, which does
+  not seem to be a problem.
+
+  This form is disallowed for:
+
+  - Type variables
+
+    ::
+
+      forall (a ::('One)). a -> Int -- rejected
+  - Top-level signatures (though, see `Toplevel binders`_)
+
+    ::
+
+      foo ::('One) A -> B -- rejected
+      foo x = …
+
+  The form is however permitted in records (see `Records`_ below)
+
+  ::
+
+    data R = R { unrestrictedField ::(Omega) A, linearField ::(One) B }
+
+In the fashion of levity polymorphism, the proposal introduces a data
+type ``Multiplicity`` which is treated specially by the type checker,
+to represent the multiplicities:
+
+- ::
+
+    data Multiplicity
+      = One    -- represents 1
+      | Omega  -- represents ω
+
+- Accompanied by two specially recognised type families:
+
+  ::
+
+    type family (:+) (p :: Multiplicity) (q :: Multiplicity) :: Multiplicity
+    type family (:*) (p :: Multiplicity) (q :: Multiplicity) :: Multiplicity
+
+  Note: unification of
+  multiplicities will be performed up to the semiring laws for
+  ``(:+)`` and ``(:*)`` (see Specification_).
+
+The linear and unrestricted arrows are aliases:
+
+- ``(->)`` is an alias for ``(:'Omega ->)``
+- ``(->.)`` (ASCII syntax) and ``(⊸)`` (Unicode syntax) are aliases
+  for ``(:'One ->)``
+
+Printing
```

Comment:
12. Even with the behavior as specified in this section, a beginner asking for `:info map` will see multiplicity polymorphism. Indeed, the same will happen for lowly `($)`, which (I assume) will get the type `(a :p-> b) ->. a :p-> b`. I suppose `:info ($)` prints levity polymorphism today, so this wouldn't be any different.

What will :info print for datatypes? Those in H98 non-record syntax are spared from difficulty, but what about those printed in GADT syntax? Does it matter whether `-XLinearTypes` is on? What if the extension is on in the module loaded but not in GHCi?

Reply:
But would that be printed without any extensions enabled? I see the `->.`. I'm not sure we want students in their first week of Haskell seeing `->.`.

In any case, I think we're saved by the specification of `:type`: `:type expr` reports the type that would be assigned to a variable if we wrote `let x = expr`. If type inference never infers a linear arrow, then we'll get `(a -> b) -> a -> b`, as desired. The `:info` output is fine, I suppose.

Reply:
I'm satisfied about the behavior of `:type` and `:info` here. But what about error messages? If, say, `Just` is a "relevant binding" in an error message, will it appear as `Just :: a :p-> Maybe a`?

Reply:
When you say `:info` or `:type` in GHCi, and in type error messages, we already have a bunch of
ad-hoc compromises that suppress kind polymorphism and levity polymorphism
unless those extensions are enables.  We could (and should) do the same for linearity
information (both polymorphsm and -o types).  It's not pretty, but it's
absolutely not new.

Reply:
This section is clear (as in, fully precise) to me, but it doesn't have the property that beginners will never see linear types. Specifically, if `map` is a "relevant binding" or a "valid hole fit" (two sections of current error messages), then its type is printed. According to this section, its type will still be `(a :p-> b) -> [a] :p-> [b]`, which I don't think we want to expose to beginners. Perhaps I've misunderstood?

Reply:
What's confusing to me, I think, is that I don't see why "let's do the same thing as type reps" is lacking in any way.

Reply:
Do you mean "runtime reps"? If so, that's not what's currently in the proposal, which describes a different mechanism.

Reply:
Right. That's what I meant. And I must admit that I don't know how the proposal differs from the treatment runtime reps. Could you enlighten me?

Reply:
With `-fprint-explicit-multiplicites` off (the default), GHC will use the abbreviations `->.` and `->` wherever possible. However, "Where no shorthand is available, as is the case for multiplicity polymorphic arrows, then the long form is used" whether or not the flag is on. This means that all constructors will be printed with multiplicity polymorphism in their types, as well as the types for linearized functions once linearity is in `base`.

The difference with `-fprint-explicit-runtime-reps` is that runtime-rep-polymorphism is still suppressed when `-fprint-explicit-runtime-reps` is not enabled, by defaulting variables to `LiftedRep`.

Reply:
Ah I see. I think that @simonpj above asserts that this would be the desirable behaviour. I haven't got time to edit the document further today. But let's assume that this is what is written until I get back to it.